README.md
YT iframe Generator
===================

_yt_iframe_ is a python module which can convert a youtube video link into an embeddable iframe.

---

## Getting started

In order to use this module, install it through your terminal.
``` console
foo@bar:~$ pip install yt-iframe
```

Import the module in Python.
``` python
from yt_iframe import yt
```

---

## Using the module

### yt.video()
>   Generates a YouTube embed video iFrame from a YouTube video link.

``` python
url = 'https://www.youtube.com/watch?v=UzIQOQGKeyI' # (Required)
width = '560' # (Optional)
height = '315' # (Optional)
iframe = yt.video(url, width=width, height=height)
```

**Parameters**
-   link : str _(required)_
    -   A link to a YouTube video.
-   width : str _(optional. default="560")_
    -   The width of the iFrame in pixels.
-   height : str _(optional. default="315")_
    -   The height of the iFrame in pixels.

**Returns**
- html : str
    -   The iFrame for the YouTube video.

---

### yt.channel()
>   Generates a list of YouTube video links from a YouTube channel.

``` python
url = 'https://www.youtube.com/user/ouramazingspace' # (Required)
videolist = yt.channel(url)
```

**Parameters**
-   link : str _(required)_
    -   A link to a YouTube channel.

**Returns**
-   links : list
    -   A list of links to YouTube videos.

---

### yt.channelDict()
>   Generates videos and metadata from a YouTube channel.

``` python
url = 'https://www.youtube.com/user/ouramazingspace'
videolist = yt.channelDict(url)

videolist['name'] # Name of channel
videolist['videos'] # Nested dictionary. Key = video title, Value = link
```

**Parameters**
-   link : str _(required)_
    -   A link to a YouTube channel.

**Returns**
-   channel : dict
    -   A dictionary of the YouTube channel's information.
    -   Key/value pairs:
        -   name = the name of the YouTube channel
        -   videos = List of video links

---

### yt.getFrames()
>   Generates a list of iFrames from a list of YouTube videos.

``` python
channel = yt.channel('https://www.youtube.com/user/ouramazingspace') # (Required)
width = '560' # (Optional)
height = '315' # (Optional)
responsive = True # (Optional)

# Fixed size iframes
iframes = yt.getFrames(channel, width=width, height=height)

# Responsive iframes
iframes = yt.getFrames(channel, responsive=responsive)
```

**Parameters**
-   links : list _(required)__
    -   A list of links to YouTube videos.
-   width : str _(optional. default="560")_
    -   The width of each iFrame in pixels.
-   height : str _(optional. default="315")_
    -   The height of each iFrame in pixels.
-   responsive : bool _(optional. default = False)_
    -   Determines whether each iFrame is dynamically or statically sized.

**Returns**
-   iframes : list
    -   List of iFrames.

---

### yt.linkResponsive()
>   Get link to css for styling the iFrames.
>   Alternatively, you can add this line of html in your head tag.
>   '<link rel="stylesheet" href="https://raw.githubusercontent.com/RobbyB97/yt-iframe-python/master/yt_iframe/yt_iframe.css">'

**Returns**
-   str
    -   HTML link tag to import css for iFrames

---

### yt.videoResponsive()
>   Generates a responsive iFrame video. Make sure you have the css file imported with the linkResponsive() function. You can wrap a bunch of these generated tags in a container and the iFrames will resize to fit the layout within that container.

``` python
url = 'https://www.youtube.com/watch?v=UzIQOQGKeyI' # (Required)
layout = 'singlecolumn' # (Optional)

video = yt.videoResponsive(url, layout=layout) # Get HTML
```

**Parameters**
-   link : str _(required)_
    -   A link to a YouTube video.
-   layout : str _(optional. default="onecolumn")_
    -   Soecifies the relative size of the iFrame.
    -   Acceptable values:
        -   'onecolumn' - Generates one column layout
        -   'twocolumn' - Generates two column video

**Returns**
-   iframes : list
    -   List of iFrames.

---

## Changelog

### == v1.0.5 ==
* _Fix css import link_
* _Fix argument error in videoResponsive()_
* _Fix xml parsing (the lxml dependency issue)_
* _Add docstrings for all functions_
* _Refactor functions to improve readability_

### == v1.0.4 ==
* _Add layout argument to videoResponsive() and getFrames()_
* _Add two column layout option to videoResponsive()_

### == v1.0.3 ==
* _Add responsive iframes_
* _getFrames() arguments changed from framewidth and frameheight to width and height_

### == v1.0.1 ==
* _Allow size of iframe to be specified in video() function_
* _Allow sizes of iframes to be specified in getFrames() function_

### == v1.0.0 ==
* _Initial release_


yt_iframe/CHANGELOG.txt
Change Log
==========

1.0.5 (11/12/20)
----------------
- Fix css import link
- Fix argument error in videoResponsive()
- Fix xml parsing (the lxml dependency issue)
- Add docstrings for all functions
- Refactor functions to improve readability

1.0.4
-----
- Add layout argument to videoResponsive() and getFrames()
- Add two column layout option to videoResponsive()

1.0.3
-----
- Add responsive iframes
- getFrames() arguments changed from framewidth and frameheight to width and height

1.0.1
-----
- Allow size of iframe to be specified in video() function
- Allow sizes of iframes to be specified in getFrames() function

1.0.0
-----
- Initial release

yt_iframe/test/__init__.py
import logging
logging.basicConfig(
    format="%%(name)-15s %(levelname)-8s %(message)s"
)



yt_iframe/test/test_video.py
from unittest import TestCase
from yt_iframe import yt


class TestVideo(TestCase):
    def setUp(self):
        self.videoUrl = "https://www.youtube.com/watch?v=dQw4w9WgXcQ"

    def test_iframe(self):
        expected = '<iframe width="560" height="315" src="https://www.youtube.com/embed/dQw4w9WgXcQ" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>'
        iframe = yt.video(self.videoUrl)
        self.assertEqual(iframe, expected)

    def test_iframe_not_valid(self):
        with self.assertRaises(yt.InvalidLink):
            yt.video(None)

    def test_iframe_no_id(self):
        with self.assertRaises(yt.InvalidLink):
            yt.video(self.videoUrl[:32])


yt_iframe/test/test_channel.py
from unittest import TestCase
from unittest.mock import patch, MagicMock
from yt_iframe import yt
from requests import Response
from os import path


class TestChannel(TestCase):
    def setUp(self):
        self.channelUrl = "https://www.youtube.com/channel/UCuAXFkgsw1L7xaCfnd5JJOw"
        basedir = path.dirname(path.realpath(__file__))
        with open(path.join(basedir, "channel_header.xml"), 'r') as header:
            self.header = header.read()
        with open(path.join(basedir, "channel_entry.xml"), 'r') as entry:
            self.entry = entry.read()
        with open(path.join(basedir, "channel_footer.xml"), 'r') as footer:
            self.footer = footer.read()



    def test_channel(self):
        expected = ['<iframe width="560" height="315" src="https://www.youtube.com/embed/dQw4w9WgXcQ" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>']

        def _get(url):
            response = MagicMock(spec=Response)
            response.text = self.header + self.entry + self.footer
            return response

        with patch('requests.get', new=_get):
            iframes = yt.channel(self.channelUrl)
            self.assertEqual(iframes, expected)

    def test_channel_invalid_url(self):
        with self.assertRaises(yt.InvalidLink):
            iframes = yt.channel(self.channelUrl[:32])

    def test_channel_invalid_feed(self):
        with patch('requests.get', side_effect=Exception()):
            with self.assertRaises(yt.InvalidFeed):
                iframes = yt.channel(self.channelUrl)

    def test_invalid_links(self):
        def _get(url):
            response = MagicMock(spec=Response)
            entry = '<entry><link rel="alternate" href="" /></entry>'
            response.text = self.header + entry + self.footer
            return response

        with patch('requests.get', new=_get):
            iframes = yt.channel(self.channelUrl)
            self.assertEqual(iframes, [])


yt_iframe/yt_iframe/yt.py
import logging
from bs4 import BeautifulSoup as bs
import requests
from time import sleep

# Exception handlers
class InvalidLink(Exception):
    pass

class InvalidFeed(Exception):
    pass


logger = logging.getLogger("yt_iframe")


# Global variables
css_link = 'https://raw.githubusercontent.com/RobbyB97/yt-iframe-python/master/yt_iframe/yt_iframe.css'


def channel(link):
    """Generates a list of YouTube video links from a given channel.
    
    Parameters
    ----------
    link : str (required)
        A link to a YouTube channel.
    
    Returns
    -------
    links : list
        A list of links to YouTube videos.
    """

    iframes = []       # list of iframes
    links = []      # list of video links

    # Inner methods for finding RSS URL
    def userURL(link):
        user = requests.get(link).text
        soup = bs(user, "html.parser")
        link = soup.find("link", {"rel":"canonical"})
        return channelURL(link['href'])

    def channelURL(link):
        try:
            link = link.split('/channel/')[1]
            if not link:
                raise InvalidLink("Link not found")

            link = 'https://www.youtube.com/feeds/videos.xml?channel_id=' + link
        except Exception as e:
            raise InvalidLink('yt.channel - Error! Not a valid link.') from e
        return link

    # Get RSS URL from channel URL
    if '/channel/' in link:
        xml = channelURL(link)
    elif '/user/' in link:
        xml = userURL(link)
    else:
        print('yt.channel - Error! Not a valid link')

    try:
        # Get RSS feed
        feed = requests.get(xml).text
        xmlsoup = bs(feed, "html.parser")
    except Exception as e:
        raise InvalidFeed('yt.channel - Error! Could not parse xml feed.') from e

    # Add video links to links list
    for entry in xmlsoup.findAll('link'):
        if '/watch?v=' in entry['href']:
            links.append(entry['href'])
    return links


def channelDict(link):
    """Generates videos and metadata from a given YouTube channel.
    
    Parameters
    ----------
    link : str (required)
        A link to a YouTube channel.
    
    Returns
    -------
    channel : dict
        A dictionary of the YouTube channel's information.
        Dictionary key/value pairs:
            name = name of the YouTube channel
            videos = List of video links
    """

    links = {}       # Key = video title, Value = video link
    channel = {}     # Master dictionary

    # Get link to RSS feed
    def userURL(link):
        user = requests.get(link).text
        soup = bs(user, "html.parser")
        link = soup.find("link", {"rel":"canonical"})
        return channelURL(link['href'])
    def channelURL(link):
        link = link.split('/channel/')[1]
        link = 'https://www.youtube.com/feeds/videos.xml?channel_id=' + link
        return link

    # Get RSS URL from channel URL
    if '/channel/' in link:
        xml = channelURL(link)
    elif '/user/' in link:
        xml = userURL(link)
    else:
        print('yt.channel - Error! Not a valid link')

    # Get RSS feed
    feed = requests.get(xml).text
    xmlsoup = bs(feed, "html.parser")

    # Get name of channel
    channel['name'] = xmlsoup.find('author').find('name').text

    # Create video dictionary entries
    for entry in xmlsoup.findAll('entry'):
        ytlink = entry.find('link')
        if '/watch?v=' in ytlink['href']:
            title = entry.find('title').text
            ytlink = ytlink['href']
            links[title] = ytlink
        else:
            continue
    channel['videos'] = links

    return channel


def video(link, width="560", height="315"):
    """Generates a YouTube embed video iFrame from a given YouTube video link.
    
    Parameters
    ----------
    link : str (required)
        A link to a YouTube video.
    width : str
        The width of the iFrame in pixels.
    height : str
        The height of the iFrame in pixels.
    
    Returns
    -------
    html : str
        The iFrame for the YouTube video.
    """

    # Ensure link is valid
    try:
        link = link.split('watch?v=')[1]
        if not link:
            raise InvalidLink("Link not found")
    except Exception as e:
        raise InvalidLink('yt.video - Error! Not a valid link.') from e
    
    # Create HTML iFrame
    html = '<iframe width="'+width+'" height="'+height+'" src="https://www.youtube.com/embed/'+link+'" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>'
    
    return html


def getFrames(links, width="560", height="315", responsive=False):
    """Generates a list of iFrames from a list of YouTube videos.
    
    Parameters
    ----------
    links : list (required)
        A list of links to YouTube videos.
    width : str
        The width of each iFrame in pixels.
    height : str
        The height of each iFrame in pixels.
    responsive : bool
        Determines whether each iframe is dynamically (true) or 
        statically (false) sized.
    
    Returns
    -------
    iframes : list
        List of iframes.
    """

    iframes = []    # List of iFrames to be returned

    for vid in links:
        # Generate iframe
        try:
            if responsive:
                frame = videoResponsive(vid, width=width, height=height)
            else:
                frame = video(vid, width=width, height=height)
            iframes.append(frame)
        except InvalidLink as e:
            logger.error(e)

    return iframes


def linkResponsive():
    """Get link to css for styling the iFrames.

    Returns
    -------
    str
        HTML link tag to import css for iFrames
    """
    return '<link rel="stylesheet" href="'+css_link+'">'


def videoResponsive(link, layout='onecolumn'):
    """Generates a responsive iFrame video. Make sure you have the css file
    imported with the linkResponsive() function. Check the README for more info.
    
    Parameters
    ----------
    link : str (required)
        A link to a YouTube video.
    layout : str
        Specifies the relative size of the iFrame. 
        Acceptable values:
            onecolumn - Generates one column layout
            twocolumn - Generates two column layout
    
    Returns
    -------
    iframes : list
        List of iframes.
    """

    # Set layout
    if layout == 'onecolumn':
        responsive_video = '<div class="yt-iframe-container">'
    elif layout == 'twocolumn':
        responsive_video = '<div class="yt-iframe-twocolumn">'
    else:
        log.warning('%s is not a proper layout. Defaulting to single column...' % layout)
        responsive_video = '<div class="yt-iframe-container">'

    # Get video and close tags
    yt_vid = video(link)
    responsive_video += yt_vid
    responsive_video += '</div>'

    return responsive_video
